kernel/kernel-api/src/main/java/io/delta/kernel/internal/icebergcompat/IcebergCompatV3MetadataValidatorAndUpdater.java

```diff
@@ -0,0 +1,198 @@
+
+import static io.delta.kernel.internal.tablefeatures.TableFeatures.*;
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+import static java.util.Collections.singletonList;
+import static java.util.stream.Collectors.toList;
+
+import io.delta.kernel.internal.DeltaErrors;
+import io.delta.kernel.internal.TableConfig;
+import io.delta.kernel.internal.actions.Metadata;
+import io.delta.kernel.internal.actions.Protocol;
+import io.delta.kernel.internal.tablefeatures.TableFeature;
+import io.delta.kernel.internal.types.TypeWideningChecker;
+import io.delta.kernel.internal.util.ColumnMapping.ColumnMappingMode;
+import io.delta.kernel.internal.util.SchemaIterable;
+import io.delta.kernel.internal.util.SchemaUtils;
+import io.delta.kernel.internal.util.Tuple2;
+import io.delta.kernel.types.*;
+import io.delta.kernel.utils.DataFileStatus;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/** Utility methods for validation and compatibility checks for Iceberg V3. */
+public class IcebergCompatV3MetadataValidatorAndUpdater
+    extends IcebergCompatMetadataValidatorAndUpdater {
+  /**
+   * Validate and update the given Iceberg V3 metadata.
+   *
+   * @param newMetadata Metadata after the current updates
+   * @param newProtocol Protocol after the current updates
+   * @return The updated metadata if the metadata is valid and updated, otherwise empty.
+   * @throws UnsupportedOperationException if the metadata is not compatible with Iceberg V3
+   *     requirements
+   */
+  public static Optional<Metadata> validateAndUpdateIcebergCompatV3Metadata(
+      boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+    return INSTANCE.validateAndUpdateMetadata(
+        new IcebergCompatInputContext(isCreatingNewTable, newMetadata, newProtocol));
+  }
+
+  /**
+   * Validate the given {@link DataFileStatus} that is being added as a {@code add} action to Delta
+   * Log. Currently, it checks that the statistics are not empty.
+   *
+   * @param dataFileStatus The {@link DataFileStatus} to validate.
+   */
+  public static void validateDataFileStatus(DataFileStatus dataFileStatus) {
+    if (!dataFileStatus.getStatistics().isPresent()) {
+      // presence of stats means always has a non-null `numRecords`
+      throw DeltaErrors.icebergCompatMissingNumRecordsStats(
+          INSTANCE.compatFeatureName(), dataFileStatus);
+    }
+  }
+
+  /// //////////////////////////////////////////////////////////////////////////////
+  /// Define the compatibility and update checks for icebergCompatV3             ///
+  /// //////////////////////////////////////////////////////////////////////////////
+
+  private static final IcebergCompatV3MetadataValidatorAndUpdater INSTANCE =
+      new IcebergCompatV3MetadataValidatorAndUpdater();
+
+  private static final IcebergCompatRequiredTablePropertyEnforcer ICEBERG_COMPAT_V3_CM_REQUIREMENT =
```

Comment:
create a follow up task to define this check in the base class and reuse it. You can post that as the first PR precuros to this PR.